resqpy/grid/grid.py

```diff
@@ -0,0 +1,2107 @@
+
+        # todo: replace with crs module calls
+
+        if crs_root is None:
+            crs_root = self.crs_root
+            if crs_root is None:
+                return
+        flat_a = a.reshape((-1, 3))  # flattened view of array a as vector of (x, y, z) points, in situ
+        x_offset = float(rqet.find_tag(crs_root, 'XOffset').text)
+        y_offset = float(rqet.find_tag(crs_root, 'YOffset').text)
+        z_offset = float(rqet.find_tag(crs_root, 'ZOffset').text)
+        areal_rotation = float(rqet.find_tag(crs_root, 'ArealRotation').text)
+        assert (areal_rotation == 0.0)
+        # todo: check resqml definition for order of rotation and translation and apply rotation if not zero
+        if global_xy_units is not None:
+            crs_xy_units_text = rqet.find_tag(crs_root, 'ProjectedUom').text
+            if crs_xy_units_text in ['ft', 'm']:  # todo: else raise exception
+                bwam.convert_lengths(flat_a[:, 0], global_xy_units, crs_xy_units_text)  # x
+                bwam.convert_lengths(flat_a[:, 1], global_xy_units, crs_xy_units_text)  # y
+        if global_z_units is not None:
+            crs_z_units_text = rqet.find_tag(crs_root, 'VerticalUom').text
+            if crs_z_units_text in ['ft', 'm']:  # todo: else raise exception
+                bwam.convert_lengths(flat_a[:, 2], global_z_units, crs_z_units_text)  # z
+        flat_a[:, 0] -= x_offset
+        flat_a[:, 1] -= y_offset
+        if z_offset != 0.0:
+            flat_a[:, 2] -= z_offset
+        if global_z_increasing_downward is not None:  # note: here negation is made in local crs; if z_offset is not zero, this might not be what is intended
+            crs_z_increasing_downward = self.z_inc_down()
+            assert crs_z_increasing_downward is not None
+            if global_z_increasing_downward != crs_z_increasing_downward:
+                negated_z = np.negative(flat_a[:, 2])
+                flat_a[:, 2] = negated_z
+
+    def write_hdf5_from_caches(self,
+                               file = None,
+                               mode = 'a',
+                               geometry = True,
+                               imported_properties = None,
+                               write_active = None,
+                               stratigraphy = True,
+                               expand_const_arrays = False):
+        """Create or append to an hdf5 file.
+        
+        Writes datasets for the grid geometry (and parent grid mapping) and properties from cached arrays.
+        """
+        # NB: when writing a new geometry, all arrays must be set up and exist as the appropriate attributes prior to calling this function
+        # if saving properties, active cell array should be added to imported_properties based on logical negation of inactive attribute
+        # xml is not created here for property objects
+
+        write_hdf5_from_caches(self, file, mode, geometry, imported_properties, write_active, stratigraphy,
```

Comment:
Here is an example of where I would prefer the called function to be prefixed with an _ to hint at privacy and also to distinguish it from the method name.